Go binding: add GetClientStatus method to Database

gm42 opened the pull request:
Add `GetClientStatus` method to `Database`.

Allow fetching client status JSON information for any database.
Added a test to make sure that it works with an open database.

By looking in `fdbclient/MultiVersionTransaction.actor.cpp` I found out that `fdb_database_get_client_status()` will return an empty string when the db is invalid; I will document this in this same PR.

Related: #11621

# Code-Reviewer Section

The general pull request guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [x] The PR has a description, explaining both the problem and the solution.
- [x] The description mentions which forms of testing were done and the testing seems reasonable.
- [x] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

johscheuer reviewed (commented):
Changes LGTM, do you think it make sense to have a go struct that can be used to deserialize the result?

gm42 commented:
> Changes LGTM, do you think it make sense to have a go struct that can be used to deserialize the result?

I was going to say that if we add it, should also be added for the machine-readable JSON status (from key `\xFF\xFF/status/json`), but I just realized that there is no dedicated Go API method to fetch it, so that's not a concern.

Pros:
* will be gladly welcome downstream
Cons:
* needs a commitment to be kept up to date every time there is a change
* if there is a breaking change in such JSON format, the Go binding will need an immediate update because otherwise nobody will be able to read any field from it (even if it's not the one affected by the breaking change)
* in case of multiple JSON formats for multiple versions, the Go binding will have to forward-port the older versions of the struct to the new version

I can give this a try, and point out fields whose type I am not sure about; what do you prefer?

gm42 commented:
@johscheuer any idea why in some cases an empty string and no error can be returned? I am trying to reproduce this, but without success so far.

gm42 commented:
Re-based and added a commit which updates the GoDoc for `ReadTransact()`.

johscheuer reviewed (approved):
LGTM 👍

gm42 commented:
@johscheuer sometimes the returned status is an empty string, it is happening also in tests:
```
--- FAIL: TestGetClientStatus (0.00s)
    fdb_test.go:402: returned status is empty
    fdb_test.go:404: client status returned:
FAIL
```

Could this be a race within libfdb?

gm42 commented:
Thanks for triggering the CI; it is still failing, I think I begin to understand what's happening.

The C client library has logic to seed an internal `versionMonitorDb` with the first working database connection which supports it, otherwise it falls back on creating one locally.

The get client status function will perform some magic and in one case can return empty string:
```C++
ThreadFuture<Standalone<StringRef>> MultiVersionDatabase::getClientStatus() {
	auto stateRef = dbState;
	auto db = stateRef->dbVar->get();
	if (!db.value.isValid()) {
		db.value = stateRef->versionMonitorDb;
	}
	if (!db.value.isValid()) {
		return onMainThread([stateRef] { return Future<Standalone<StringRef>>(stateRef->getClientStatus(""_sr)); });
	} else {
```

So I believe that for some reason neither databases are considered valid.

gm42 commented:
> So I believe that for some reason neither databases are considered valid.

The reason turned out to be: multi-version client must be enabled in order for this function call to return non-empty string.

johscheuer reviewed (approved):
LGTM 👍

gm42 commented:
@jzhou77 I think it needs a close/open cycle to trigger CI?

jzhou77 commented:
ctest has a failure:
```
55/72 Test #70: fdb_go_test .........................................................***Failed    3.63 sec
Database created
setOne called with:  fdb.Database
getOne called with: fdb.Database
--- FAIL: TestGetClientStatusUnavailable (0.00s)
    fdb_test.go:400: could not disable multi-version client: FoundationDB error code 2007 (Option not valid in this context)
Assertion !networkStartSetup failed @ /codebuild/output/src83525386/src/github.com/apple/foundationdb/fdbclient/MultiVersionTransaction.actor.cpp 3021:
  addr2line -e libfdb_c.so-debug -p -C -f -i 0x13ab59c 0x93f7ae 0x8161bb 0xffff80e2f0d06a27 0xffff80e2f0c5b804
--- FAIL: TestGetClientStatus (0.00s)
    fdb_test.go:416: FoundationDB error code 4100 (An internal error occurred)
FAIL

```

gm42 commented:
Thanks, I have removed that test; it can be added back if we figure out how to run with multi-version-client disabled.

gm42 commented:
By the way, that `4100` error is something I also spotted [here](https://github.com/apple/foundationdb/issues/11621#issuecomment-2331514094); do we need a separate issue for this?

gm42 commented:
@jzhou77 can the CI be triggered again please?

gm42 commented:
@jzhou77 I rebased it, it needs another CI run

gm42 commented:
Thanks for re-triggering the CI; I can check next Friday why it's failing and update this PR.

gm42 commented:
@jzhou77 the error seems to be about code formatting (see below); however I have no idea how to address this :thinking: is it due to something I changed in this PR? I have also just re-based the PR, in case it helps.

```
CODE FORMAT NOT CLEAN

THE FOLLOWING FILES NEED TO BE FORMATTED

bindings/c/foundationdb/fdb_c.h
bindings/c/test/apitester/TesterTransactionExecutor.cpp
bindings/c/test/mako/admin_server.hpp
bindings/c/test/test.h
bindings/c/test/unit/trace_partial_file_suffix_test.cpp
bindings/flow/IDirectory.h
bindings/flow/fdb_flow.h
contrib/SimpleOpt/include/SimpleOpt/SimpleOpt.h
contrib/md5/md5.c
contrib/rapidjson/rapidjson/document.h
contrib/rapidjson/rapidjson/rapidjson.h
contrib/rapidjson/rapidjson/reader.h
contrib/rapidjson/rapidjson/writer.h
contrib/rapidxml/include/rapidxml/rapidxml.hpp
contrib/sqlite/sqlite3.h
contrib/sqlite/sqliteInt.h
fdbclient/BlobGranuleReader.actor.cpp
fdbclient/FileBackupAgent.actor.cpp
fdbclient/NativeAPI.actor.cpp
fdbclient/include/fdbclient/Atomic.h
fdbclient/include/fdbclient/BackupAgent.actor.h
fdbclient/include/fdbclient/BlobGranuleRequest.actor.h
fdbclient/include/fdbclient/IConfigTransaction.h
fdbclient/include/fdbclient/RandomKeyValueUtils.h
fdbclient/include/fdbclient/SpecialKeySpace.actor.h
fdbclient/include/fdbclient/StorageServerInterface.h
fdbclient/include/fdbclient/TaskBucket.h
fdbclient/include/fdbclient/TransactionLineage.h
fdbclient/sha1/SHA1.cpp
fdbrpc/FlowTests.actor.cpp
fdbrpc/include/fdbrpc/AsyncFileKAIO.actor.h
fdbrpc/include/fdbrpc/LoadBalance.actor.h
fdbrpc/include/fdbrpc/MultiInterface.h
fdbrpc/include/fdbrpc/ReplicationPolicy.h
fdbrpc/libeio/ecb.h
fdbserver/ConfigDatabaseUnitTests.actor.cpp
fdbserver/DDRelocationQueue.actor.cpp
fdbserver/DiskQueue.actor.cpp
fdbserver/KeyValueStoreRocksDB.actor.cpp
fdbserver/KeyValueStoreSQLite.actor.cpp
fdbserver/KeyValueStoreShardedRocksDB.actor.cpp
fdbserver/RocksDBCheckpointUtils.actor.cpp
fdbserver/VersionedBTree.actor.cpp
fdbserver/include/fdbserver/BulkDumpUtil.actor.h
fdbserver/include/fdbserver/CoordinationInterface.h
fdbserver/include/fdbserver/DDTxnProcessor.h
fdbserver/include/fdbserver/DataDistribution.actor.h
fdbserver/include/fdbserver/RemoteIKeyValueStore.actor.h
fdbserver/include/fdbserver/RestoreCommon.actor.h
fdbserver/include/fdbserver/RestoreController.actor.h
fdbserver/include/fdbserver/RestoreLoader.actor.h
fdbserver/include/fdbserver/RestoreRoleCommon.actor.h
fdbserver/include/fdbserver/RestoreWorkerInterface.actor.h
fdbserver/include/fdbserver/RoleLineage.actor.h
fdbserver/include/fdbserver/TestTLogServer.actor.h
fdbserver/include/fdbserver/workloads/workloads.actor.h
fdbserver/workloads/AsyncFileRead.actor.cpp
fdbserver/workloads/PrivateEndpoints.actor.cpp
fdbservice/FDBService.cpp
flow/Net2.actor.cpp
flow/Platform.actor.cpp
flow/Trace.cpp
flow/include/flow/Buggify.h
flow/include/flow/CodeProbe.h
flow/include/flow/Error.h
flow/include/flow/IndexedSet.h
flow/include/flow/Optional.h
flow/include/flow/TDMetric.actor.h
flow/include/flow/ThreadHelper.actor.h
flow/include/flow/ThreadPrimitives.h
flow/include/flow/Trace.h
flow/include/flow/Traceable.h
flow/include/flow/WipedString.h
flow/include/flow/ppc-opcode.h
flow/include/flow/singleton.h
flow/include/flow/sse2neon.h
flow/include/flow/xxhash.h
```

jzhou77 commented:
code format error should be only for ARM build, because we've changed build environment to clang 19. You can ignore these errors.

gm42 commented:
Thanks; now the failure I see is a failed assertion:
```
56/73 Test #71: fdb_go_test .........................................................***Failed    3.66 sec
Database created
setOne called with:  fdb.Database
getOne called with: fdb.Database
Assertion !networkStartSetup failed @ /codebuild/output/src2081500999/src/github.com/apple/foundationdb/fdbclient/MultiVersionTransaction.actor.cpp 3021:
  addr2line -e libfdb_c.so-debug -p -C -f -i 0x13d06ac 0x94b7ae 0x8221bb 0xffff8064b439fd07 0xffff8064b42f4804
--- FAIL: TestGetClientStatus (0.00s)
    fdb_test.go:397: FoundationDB error code 4100 (An internal error occurred)
FAIL
```

`libfdb_c.so-debug` is not available in `workspace.zip`, by using the available one I can symbolize as:
```
$ addr2line -e build_output/lib/libfdb_c.so -p -C -f -i 0x13d06ac 0x94b7ae 0x8221bb 0xffff8064b439fd07 0xffff8064b42f4804
internal_error_impl(char const*, char const*, int) at /usr/local/bin/../include/c++/v1/string:1499
 (inlined by) __get_pointer at /usr/local/bin/../include/c++/v1/string:1620
 (inlined by) data at /usr/local/bin/../include/c++/v1/string:1279
 (inlined by) c_str at /usr/local/bin/../include/c++/v1/string:1277
 (inlined by) internal_error_impl at /codebuild/output/src2081500999/src/github.com/apple/foundationdb/flow/Error.cpp:64
MultiVersionApi::setNetworkOptionInternal(FDBNetworkOptions::Option, Optional<StringRef>) at /codebuild/output/src2081500999/src/github.com/apple/foundationdb/fdbclient/MultiVersionTransaction.actor.cpp:?
fdb_network_set_option at /codebuild/output/src2081500999/src/github.com/apple/foundationdb/bindings/c/fdb_c.cpp:154
?? ??:0
?? ??:0
```

So my conclusion is that setting network option causes a 4100 internal error; I removed that from the test and moved it to a new example.

gm42 commented:
@jzhou77 can you please open/close to trigger CI? Tests should pass now

gm42 commented:
@johscheuer the test fails because multi-version client is not available:
```
--- FAIL: TestGetClientStatus (0.00s)
    fdb_test.go:399: GetClientStatus failed multi-version client API is unavailable
Assertion !networkStartSetup failed @ /codebuild/output/src3661653684/src/github.com/apple/foundationdb/fdbclient/MultiVersionTransaction.actor.cpp 3021:
  addr2line -e libfdb_c.so-debug -p -C -f -i 0x13d04ec 0x94b7ae 0x8221bb 0xffff80c5f664af67 0xffff80c5f659f804
```

What do you advise here? Should I remove the test or there is a way in tests to have the multi-version client enabled?

gm42 commented:
Thanks!
